### ISSUE TITLE
Google Cloud DNS - Vulnerable 

### ISSUE THREAD

**Issue by libertalialtd** [indianajson/can-i-take-over-dns#2]
### <code>Nameserver</code>

ns-cloud-**.googledomains.com

### <code>Explanation</code>

If a domain points to one of the nameservers listed above it is using Google Cloud DNS, a free service. A `SERVFAIL` error indicates the domain is vulnerable to take over. To perform the takeover <a href="console.cloud.google.com">set up a free Google Cloud</a> account then navigate to <a href="https://console.cloud.google.com/net-services/dns/zones">Cloud DNS</a>. Click `Create Zone` and then enter the (sub)domain name in the field named `DNS name`. Your new zone will be given four random Google nameservers. These must match the ones on the vulnerable domain. If they do not match simply delete the zone and create another one, you should be assigned a different random set of Google nameservers. I can take a few attempts to get them to match.

**Comment by rew1nter** [comment]
Is reaching the a* ns server impossible now?
I've tried about 70 times. Only getting b* - e*.

**Comment by indianajson** [comment]
@molitona I am not sure about this, though the "GC" stands for "Google Cloud" and it may be only for internal testing.

**Comment by b1bek** [comment]
Can anyone confirm if this still works? I am getting an error saying to verify the domain ownership.
![image](https://user-images.githubusercontent.com/43294620/221164774-357ace48-561a-4b82-a6b9-8dc3375f614f.png)

**Comment by mheranco** [comment]
@b1bek  This means the parent domain is already taken, so all subdomains of this domain should not be vulnerable.

I just found a subdomain, where the zone can be created. But I also have the problem that I do not get the needed letter "c" in the NS servers. So I think Google has some mitigation in place similar to AWS, where you cannot get the same NS, when they were previously used.

**Comment by KarahanGuner** [comment]
> @b1bek This means the parent domain is already taken, so all subdomains of this domain should not be vulnerable.
> 
> I just found a subdomain, where the zone can be created. But I also have the problem that I do not get the needed letter "c" in the NS servers. So I think Google has some mitigation in place similar to AWS, where you cannot get the same NS, when they were previously used.

Same problem here. I tried like 25 times and didnt get the letter "c". Is there a workaround for this problem?

**Comment by indianajson** [comment]
I'm going to do some testing soon on this and see what I can figure out. Thanks for all the comments.

**Comment by freep0int** [comment]
is it still working?

**Comment by b1bek** [comment]
Might be patched. No any lucks in past few months.

**Comment by RogueSMG** [comment]
Seems fixed...

![gcp-fixed](https://github.com/indianajson/can-i-take-over-dns/assets/41023447/0960d242-ebe9-4e8a-a76a-c83e3696dd44)

**Comment by b1bek** [comment]
Yeah, which tool are you using btw?

**Comment by RogueSMG** [comment]
Ah that's just a "hacky" bash script I created which uses Gcloud CLI @b1bek .

Just pushed it on Github you can check it out:
https://github.com/RogueSMG/GCPShardJuggler

**Comment by testert1ng** [comment]
> Ah that's just a "hacky" bash script I created which uses Gcloud CLI @b1bek .
> 
> Just pushed it on Github you can check it out: https://github.com/RogueSMG/GCPShardJuggler

Thanks for the tool, just confirmed it is still vulnerable. Got the expected shards after ~70 tries

**Comment by indianajson** [comment]
> Thanks for the tool @RogueSMG , just confirmed it is still vulnerable. Got the expected shards after ~70 tries

Thanks for verifying this is still vulnerable as of July 2023.

**Comment by GKNSB** [comment]
I was trying to hit e1 shards. After more than 4000 attempts I never got them, only a1, b1, c1, d1. I'll keep looking into it.

**Comment by dadsgone0** [comment]
Are you sure this still works? Google dns doesn't let you delete a zone unless you delete all the record in it first. Maybe im just being dumb here but idk.

**Comment by levinwinter** [comment]
From the docs: https://cloud.google.com/dns/quotas#name-server-limits
Doesn't this effectively make DNS takeover on GCP an edge-case?

**Comment by yeti-code** [comment]
> From the docs: https://cloud.google.com/dns/quotas#name-server-limits
> Doesn't this effectively make DNS takeover on GCP an edge-case?

Yes, most GCP takeovers are edge-cases. It's also possible "in theory" to take over. subdomain via dangling A records if the DNS record is pointing to an ephemerally assigned GCP owned IP address. Once that IP gets released back into the GCP ephemeral IP pool, you "in theory" can take it over by generating VM instances over and over until you get assigned the IP that is still in the targets DNS panel in the Google Cloud Platform management panel. It's very difficult. I wrote some scripts to do this. But I have yet to "prove" that it actually works.


https://github.com/yeti-code/GCP-Genie

**Comment by DeathsPirate** [comment]
> > From the docs: https://cloud.google.com/dns/quotas#name-server-limits
> > Doesn't this effectively make DNS takeover on GCP an edge-case?
> 
> Yes, most GCP takeovers are edge-cases. It's also possible "in theory" to take over. subdomain via dangling A records if the DNS record is pointing to an ephemerally assigned GCP owned IP address. Once that IP gets released back into the GCP ephemeral IP pool, you "in theory" can take it over by generating VM instances over and over until you get assigned the IP that is still in the targets DNS panel in the Google Cloud Platform management panel. It's very difficult. I wrote some scripts to do this. But I have yet to "prove" that it actually works.
> 
> https://github.com/yeti-code/GCP-Genie

Interesting, I've just been doing the same in AWS but I'm approaching it from the angle of looking what traffic I get for an Elastic IP and seeing if the host header is valid domain then working backwards from that.   Happy to collab if you are still working on this.

**Comment by indianajson** [comment]
Given the changes we have seen I'm updating this to edge case here and in the list.

**Comment by yeti-code** [comment]
> > > From the docs: https://cloud.google.com/dns/quotas#name-server-limits
> 
> > > Doesn't this effectively make DNS takeover on GCP an edge-case?
> 
> > 
> 
> > Yes, most GCP takeovers are edge-cases. It's also possible "in theory" to take over. subdomain via dangling A records if the DNS record is pointing to an ephemerally assigned GCP owned IP address. Once that IP gets released back into the GCP ephemeral IP pool, you "in theory" can take it over by generating VM instances over and over until you get assigned the IP that is still in the targets DNS panel in the Google Cloud Platform management panel. It's very difficult. I wrote some scripts to do this. But I have yet to "prove" that it actually works.
> 
> > 
> 
> > https://github.com/yeti-code/GCP-Genie
> 
> 
> 
> Interesting, I've just been doing the same in AWS but I'm approaching it from the angle of looking what traffic I get for an Elastic IP and seeing if the host header is valid domain then working backwards from that.   Happy to collab if you are still working on this. 
> 
> 

Same. I recently updated my script to work backwards. Grab an IP and check against all targets on bugcrowd. But I still have issues since GCP temporarily "assigns" an IP from the ephemeral pool to your GCP project. So if you delete the vm and build a new one, (in this case lets say you have 4 being generated at once) you just get the same IP being rotated among them. I know that a well known group or "person" on bugcrowd called d0xing has figured out a reliable way to grab subdomains with A records / IP's owned by google. But I still have not been successful in replicating the automation. Either I am doing something wrong or it is a "scaling" / timing issue.

**Comment by indianajson** [comment]
@yeti-code Now, if we could just figure out how automate DNS takeovers on Cloudflare we'd all be rich.
